### ISSUE TITLE
Possible regression for programmatic use in `j`

### ISSUE THREAD

**Issue by jangorecki** [Rdatatable/data.table#2816]
Reported as comment in https://github.com/Rdatatable/data.table/issues/2655#issuecomment-385568576
@jjyyyin

> I don't know if this is addressed before but how do you make sure the new .. prefix works with the assignment operator := as well? 
> Consider the following example:
> `dt <- data.table(x=rnorm(10), y=runif(10));
> dt[, x := x + 1]`
> It is totally acceptable. If I have to find x programmatically, then I would do this before
> `a <- "x";
> dt[, eval(as.name(a)) := eval(as.name(a)) + 1]`.
> It worked perfectly fine until I updated the data.table package to v1.10.4.
> Now I cannot do 
> `dt[, z:=..a]`
> or 
> `dt[, ..a := ..a + 1]`
> or even 
> `dt[, eval(as.name(a)) := eval(as.name(a)) + 1]` anymore. What is the point of the ..prefix if you cannot do assignment operations with it? Anyone knows what caused the old eval(as.name(VAR)) to break down? Thanks.

I tested your examples against 1.10.2, 1.10.4 and recent 1.10.5 (upcoming 1.11.0). Your code didn't work on neither of them.
> It worked perfectly fine until I updated the data.table package to v1.10.4.

You did not specified from which version you did upgrade to 1.10.4, so hard to know which one to test.

Below results of run for three mentioned versions, please check if I haven't simplified your code too much. I also added one last call which should work on any data.table version, thus may be useful if you are developing a package having stated dependency on old data.table version.
```r
## 1.10.2
library(data.table)
dt = data.table(x=1:3)
dt[, x := x + 1]
var = "x"
dt[, eval(as.name(var)) := eval(as.name(var)) + 1]
#Error in eval(as.name(var)) : object 'x' not found
dt[, newcol:=..var]
#Error in eval(jsub, SDenv, parent.frame()) : object '..var' not found
dt[, ..var := ..var + 1]
#Error in eval(jsub, SDenv, parent.frame()) : object '..var' not found
dt[, eval(as.name(var)) := eval(as.name(var)) + 1]
#Error in eval(as.name(var)) : object 'x' not found

dt[, eval(call(":=", as.name(var), call("+", as.name(var), 1)))]
```

```r
## 1.10.4
library(data.table)
dt = data.table(x=1:3)
dt[, x := x + 1]
var = "x"
dt[, eval(as.name(var)) := eval(as.name(var)) + 1]
#Error in eval(as.name(var)) : object 'x' not found
dt[, newcol:=..var]
#Error in eval(jsub, SDenv, parent.frame()) : object '..var' not found
dt[, ..var := ..var + 1]
#Error in eval(jsub, SDenv, parent.frame()) : object '..var' not found
dt[, eval(as.name(var)) := eval(as.name(var)) + 1]
#Error in eval(as.name(var)) : object 'x' not found

dt[, eval(call(":=", as.name(var), call("+", as.name(var), 1)))]
```

```r
## upcoming 1.11.0
library(data.table)
dt = data.table(x=1:3)
dt[, x := x + 1]
var = "x"
dt[, eval(as.name(var)) := eval(as.name(var)) + 1]
#Error in eval(as.name(var)) : object 'x' not found
dt[, newcol:=..var]
dt
#   x newcol
#1: 2      x
#2: 3      x
#3: 4      x
dt[, ..var := ..var + 1]
#Error in ..var + 1 : non-numeric argument to binary operator
dt[, eval(as.name(var)) := eval(as.name(var)) + 1]
#Error in eval(as.name(var)) : object 'x' not found

dt[, eval(call(":=", as.name(var), call("+", as.name(var), 1)))]
```
So 1.10.4 is consistent to 1.10.2, while recent 1.10.5 which adds supports for `..` prefix process one of the calls instead of raising error.

**Comment by mattdowle** [comment]
Thanks @jangorecki  and @jjyyyin.   More work to do then, but doesn't look like there's an actual regression in that something that worked before no longer works.  Postponing to 1.11.2.

**Comment by jjyyyin** [comment]
Thanks for the investigation. Sorry forgot to mention that my previous version was 1.9.6. 

Another thing with the ..prefix is that you can do 
`dt[, ..var := 1]` now and get a new column with name "..var".  This bothers me as there is ambiguity when you try to access `dt[, ..var]`.

**Comment by jjyyyin** [comment]
Since you are working on this, a feature request would be to make this work when calling the assignment operator similar to a function `dt[, ":="(..var = ..var + 1)]`.

This doesn't work right now which makes the code very verbose if I want to update more columns dynamically. Thanks.

**Comment by MichaelChirico** [comment]
as ever, symbol on LHS of := will need to be wrapped in ()

**Comment by mattdowle** [comment]
Thanks all.  Looks like no regression then,  but more work needed on `..` prefix (new and expanding feature) together with `:=`. Postponing to 1.11.4.  (1.11.2 is a quick clear-up release and pressing.)

**Comment by AnonymousBoba** [comment]
@jjyyyin `dt[, ..a := ..a + 1]` (https://github.com/Rdatatable/data.table/issues/2655#issuecomment-385568576) seems to be using the `..` (dot dot) prefix in two different senses: (a) the character vector given by `a` in parent scope (accepted by LHS of `:=`) and (b) the column whose name is given by the character vector in `a` (similar to `get(..a)`) (accepted by RHS). The use of (a) is unsupported (even when wrapped with `()`) but is essentially what `..` means already. The latter sense (b) of `..` in `..a + 1` would be an entirely new meaning and somewhat similar to `@` in https://github.com/Rdatatable/data.table/issues/2655#issuecomment-375849897. Currently, `dt[, ..a]` only gives the values of the corresponding column because it behaves as `with = FALSE` when the `j` field consists of column names only; it does not yield the values of the column in and of itself (e.g., when `j` is an expression that is not just column names or indices).

If I may, I don't think it would be a good idea to add meaning (b). Of the two, only (a) is consistent with the current meaning of `..` in a `j` expression. Expanding `..` so it can work more universally in different contexts sounds great (e.g,, in `i`) but adding new definitions that may clash seems somewhat counter to its purpose (disambiguating variables).

**Comment by ColeMiller1** [comment]
What is the FR? #4163 would provide some help:

```
library(data.table)
dt = data.table(x=1:3)
var = "x"
dt[, names(.SD) := lapply(.SD, `+`, 1), .SDcols = var]
```
Otherwise, it is definitely doable to allow ```..var := x + 1``` without parentheses. But I am not sure we would want two ways to do the same thing (i.e., the other way is ```(var) := x + 1```).  But line 1038 is what we would need to modify to allow ```..var``` to be OK on the LHS.

https://github.com/Rdatatable/data.table/blob/b1b1832b0d2d4032b46477d9fe6efb15006664f4/R/data.table.R#L1032-L1042

Proof of concept:
```
var = "x"
lhs = quote(..var)
lhs_char = as.character(lhs)
if (substring(lhs_char, 1, 2) == "..") {
  lhs = get(substring(lhs_char, 3))
} else {
  lhs = lhs_char
}
lhs
[1] "x"
```

**Comment by jangorecki** [comment]
#4304 proposed new API for that
```r
dt = data.table(x=1:3)
var = "x"
dt[, var := var+1L, env=list(var="x")]
```

**Comment by ColeMiller1** [comment]
Regarding the proposed API, is the second line needed? The line is ```var = "x"```

**Comment by msummersgill** [comment]
Is there any outstanding action to be done on this issue? Looking through the issues on milestone 1.14.9 to see if there is anywhere I may be able to help, and it's not clear where this one stands. 

If stack overflow and github supported regex search it would be easier to determine how widely LHS unwrapped `..var` usage might be, but that is not the case. _(I checked my own files with `\[.+\.{2}[a-Z].+]` and came across no matches, and about 50 uses on as documented on RHS)_ 

At any rate, it seems like the changes proposed by @ColeMiller1 **are not** necessary to avoid causing a breaking change between current CRAN release, and could potentially be pushed to a future milestone consensus deems the expanded`..var` support or `names(.SD)` constructs to be worth adding.

Going forward, I see the `env()` approach to programming as being the best clear path forward to avoid a confusing variety of of `get()`, `mget()`, `eval(parse(...))`, `substitute()` and `..var` required depending on the task at hand, and look forward to it being included on the CRAN release!

**Comment by jangorecki** [comment]
As summarized by Matt

> Looks like no regression then, but more work needed on `..` prefix (new and expanding feature) together with `:=`. Postponing to 1.11.4. (1.11.2 is a quick clear-up release and pressing.)

No regression, `..var` is not yet well plugged in with `:=`. For the current moment please use newly added env var, which cover well your presented use case. Therefore use case requested could be said to be resolved by #4304.
If you still wish to have `..var` and `:=` playing together rather than env var, please submit minimal example as new issue. This one is unnecessarily complex and takes too much time to follow.
